phoenix-core/src/it/java/org/apache/phoenix/end2end/ParameterizedIndexUpgradeToolIT.java

```diff
@@ -0,0 +1,330 @@
+    private Admin admin;
+    private IndexUpgradeTool iut;
+
+    @Before
+    public void setup () throws Exception {
+        optionsBuilder = new StringBuilder();
+
+        setClusterProperties();
+
+        setUpTestDriver(new ReadOnlyProps(serverProps.entrySet().iterator()),
+                new ReadOnlyProps(clientProps.entrySet().iterator()));
+
+        conn = DriverManager.getConnection(getUrl(), new Properties());
+        conn.setAutoCommit(true);
+        ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                .getQueryServices();
+        admin = queryServices.getAdmin();
+        iut = new IndexUpgradeTool(upgrade ? UPGRADE_OP : ROLLBACK_OP, INPUT_LIST,
+                null, "/tmp/index_upgrade_" + UUID.randomUUID().toString(),true,
+                Mockito.mock(IndexTool.class));
+        iut.setConf(getUtility().getConfiguration());
+        iut.setTest(true);
+        if (!mutable) {
+            optionsBuilder.append(" IMMUTABLE_ROWS=true");
+        }
+        tableDDLOptions = optionsBuilder.toString();
+        prepareSetup();
+    }
+
+    private void setClusterProperties() {
+
+        // we need to destroy the cluster if it was initiated using property as true
+        if (Boolean.toString(upgrade).equals(clientProps
+                .get(QueryServices.INDEX_REGION_OBSERVER_ENABLED_ATTRIB))
+                || Boolean.toString(!isNamespaceEnabled).equals(serverProps
+                .get(QueryServices.IS_NAMESPACE_MAPPING_ENABLED))) {
+            tearDownMiniClusterAsync(1);
+        }
+        //setting up properties for namespace
+        clientProps.put(QueryServices.IS_NAMESPACE_MAPPING_ENABLED,
+                Boolean.toString(isNamespaceEnabled));
+        clientProps.put(QueryServices.IS_SYSTEM_TABLE_MAPPED_TO_NAMESPACE,
+                Boolean.toString(isNamespaceEnabled));
+        serverProps.putAll(clientProps);
+        //To mimic the upgrade/rollback scenario, so that table creation uses old/new design
+        clientProps.put(QueryServices.INDEX_REGION_OBSERVER_ENABLED_ATTRIB,
+                Boolean.toString(!upgrade));
+    }
+
+    private void prepareSetup() throws SQLException {
+        //inputList is "TEST.MOCK1,TEST1.MOCK2,TEST.MOCK3";
+        if (isNamespaceEnabled) {
+            conn.createStatement().execute("CREATE SCHEMA TEST");
+            conn.createStatement().execute("CREATE SCHEMA TEST1");
+        }
+        conn.createStatement().execute("CREATE TABLE TEST.MOCK1 (id bigint NOT NULL "
+                + "PRIMARY KEY, a.name varchar, sal bigint, address varchar)"+tableDDLOptions);
+        conn.createStatement().execute("CREATE TABLE TEST1.MOCK2 (id bigint NOT NULL "
+                + "PRIMARY KEY, name varchar, city varchar, phone bigint)"+tableDDLOptions);
+        conn.createStatement().execute("CREATE TABLE TEST.MOCK3 (id bigint NOT NULL "
+                + "PRIMARY KEY, name varchar, age bigint)"+tableDDLOptions);
+
+        //views
```

Comment:
nit: please mark "disabled until after PHOENIX-5385"

Reply:
I have added the global comment at the beginning of the test class. Should it be repeated?

Reply:
Ah, I missed that line. Since other people might miss it too, maybe move it to near the affected comments, rather than at the top?